Added Internet Archive Audio API provider script

amartya-dev opened the pull request:
<!-- Please replace #XX below with an existing issue number. Remove the line entirely if none exist. -->
Fixes #319 

This is an attempt to write the provider script for the Internet Archive. 

## Description
<!-- Add your description below. -->
The script added gets the details for the audio files from the Internet Archive API using the query mediatype:audio and uses the described points in the issue. The following is the information about the API: 

- [x] Verify there is a way to retrieve the entire relevant portion of the provider's collection in a systematic way via their API.
Yes the files can be fetched systematically after deciding the number of entries that should be included in a page. The API also provides a way for pagination as we can provide the parameter page in the API request. 
The endpoint is: https://archive.org/advancedsearch.php.
Documentation for the API: https://blog.archive.org/developers/
The other official documentation:  https://archive.org/services/docs/api/
The other documentation provides for a command line script and a python wrapper which can be used after obtaining the API credentials from Internet Archive.

- [x] Verify the API provides license info (license type and version; license URL provides both, and is preferred)
The API provides the license URL with the licenseurl key in the response JSON. 

- [x] Verify the API provides stable direct links to individual works.
The API does not provide the links directly but they can be easily formed by the identifier and metadata provided by querying a separate endpoint. 

- [x] Verify the API provides a stable landing page URL to individual works.
The API provides a stable landing page URL for the works: https://archive.org/details/<identifier>

- [x] Note other info the API provides, such as thumbnails, dimensions, attribution info (required if non-CC0 licenses will be kept), title, description, other meta data, tags, etc.

- [x] Attach example responses to API queries that have the relevant info.
Example response: 
{'responseHeader': {'status': 0,
  'QTime': 788,
  'params': {'query': 'mediatype:audio',
   'qin': 'mediatype:audio',
   'fields': 'identifier,title,mediatype,collection,licenseurl,date',
   'wt': 'json',
   'rows': '2',
   'start': 0}},
 'response': {'numFound': 9441652,
  'start': 0,
  'docs': [{'collection': ['audio_sermons', 'audio_religion'],
    'identifier': 'JesusTheRescuer',
    'licenseurl': 'http://creativecommons.org/licenses/by-nc-nd/3.0/',
    'mediatype': 'audio',
    'title': 'Jesus, The Rescuer'},
   {'collection': ['audio_sermons', 'audio_religion'],
    'date': '2015-07-05T00:00:00Z',
    'identifier': 'July52015EveningSermon',
    'mediatype': 'audio',
    'title': 'What to Do When the Foundations Are Destroyed'}]}}

## The following recommendations were followed
- [x] The script should be in the src/cc_catalog_airflow/dags/provider_api_scripts/ directory.
- [x] The script must use the ImageStore class (Import this from
src/cc_catalog_airflow/dags/provider_api_scripts/common/storage/image.py).
- [ ] The script should have a test suite in the same directory. (to be done later) 
- [x] The script should use the DelayedRequester class (Import this from
src/cc_catalog_airflow/dags/provider_api_scripts/common/requester.py).
- [x] The script must not use anything from
src/cc_catalog_airflow/dags/provider_api_scripts/modules/etlMods.py, since
that module is deprecated.
- [x] If the provider API has can be queried by 'upload date' or something similar,
the script should take a --date parameter when run as a script, giving the
date for which we should collect images. The form should be YYYY-MM-DD (so,
the script can be run via python my_favorite_provider.py --date 2018-01-01).
- [x] The script must provide a main function that takes the same parameters as from
the CLI. In our example from above, we'd then have a main function
my_favorite_provider.main(date). The main should do the same thing calling
from the CLI would do.
- [x] The script must conform to PEP8. Please use pycodestyle (available via
pip install pycodestyle) to check for compliance.
- [x] The script should use small, testable functions.
- [x] The test suite for the script may break PEP8 rules regarding long lines where
appropriate (e.g., long strings for testing).

## Other information
<!-- Add any other information below or delete the "Other Information" line entirely. -->
The present script is provided towards application for GSOC 2020.

## Checklist
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository.
- [x] My commit messages follow [best practices][best_practices].
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

<!-- Make sure you read and understand the following attestation. -->
## Developer Certificate of Origin
```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

amartya-dev committed:
Added Internet Archive Audio API provider script

amartya-dev committed:
Attempts for multiprocessing for different threads for each page : (experimental)

mathemancer reviewed (changes requested):
Overall, looking pretty good.  Mostly clean style, and easy to read.  Thanks for that!

Please make sure to remove all unused imports, constants, and variables.

Also, I'm curious if you tried this both with and without multiprocessing.  I'd be a bit surprised if you're getting any benefit there in actual running time.  Have you verified whether there are any rate limits for the IA API?  We should be careful about that.  Please consider the following:

- [ ] Check whether multiprocessing gives any practical benefit, and if so whether it's significant enough to warrant increased complexity.
- [ ] Remove unused imports, constants, and variables.
- [ ] Please copy the info you've gathered about the API to the Issue #319
- [ ] Please add a test suite.

Thank you!

amartya-dev committed:
Removed unused imports and variables

amartya-dev committed:
Added test suite

amartya-dev commented:
- [x] Check whether multiprocessing gives any practical benefit, and if so whether it's significant enough to warrant increased complexity.

Without multiprocessing: (Run for 2020-03-17)
2020-03-19 10:26:50,439
2020-03-19 10:33:34,809
6 minutes and 44 seconds

With multiprocessing: (Run for 2020-03-17)
2020-03-19 10:18:41,018
2020-03-19 10:25:07,873
6 minutes 26 seconds

It does give a good time difference considering the number of pages processed were only two, since the code would be dividing different pages amongst different threads, the time difference will be significant for more number of pages.

- [x]  Remove unused imports, constants, and variables.

I have removed any unused imports in the latest commit.

- [x]  Please copy the info you've gathered about the API to the Issue #319

- [x]  Please add a test suite.

Added in the latest commit

mathemancer reviewed (changes requested):
Overall, the code is clean and easy to read.  Thanks for that again!

However, the tests aren't running.  There are two reasons:
1. There's a syntax error
2. The tests call out to 3rd party APIs.  This is not a good idea, and is leading to some tests failing.  The ones that do this but are currently passing are probably on borrowed time, as well.

Thank you for providing some data on the multiprocessing gains.  I think, however, that they show that the gains are not significant enough to use the tool.  Even in the case where the function was used on a large number of pages, the current design means you'd have to hold all gathered data in memory before processing it.  This is untenable for large amounts of data.

Luckily, there won't likely be large numbers of pages, since we're restricting our attention to one day at a time.  However, this means multiprocessing doesn't really gain anything.  

Instead, please try to come up with an implementation that

1. pulls info for a single page
1. processes that page
1. writes the data to disk at some reasonable interval of rows
1. goes back to step 1 (for the next page)

This will save on memory, be easier to test and debug, and be more robust, because if something goes wrong with the network, we'll already have written as much to disk as possible, instead of losing everything.

amartya-dev committed:
Modified test to mock api calls
Removed multi processing
Commit to disk after 100 lines per page

amartya-dev committed:
Modified tests according to function in script

amartya-dev commented on src/cc_catalog_airflow/dags/provider_api_scripts/test_internet_archive.py:
```diff
@@ -0,0 +1,185 @@
+
+    actual_meta = ia._get_meta_data(identifier)
+
+    assert expected_meta == actual_meta
```
I have changed the test in order to Mock API calls and use them for unit tests. Now the test is passing

amartya-dev commented on src/cc_catalog_airflow/dags/provider_api_scripts/test_internet_archive.py:
```diff
@@ -0,0 +1,185 @@
+    end_timestamp = '2019-03-19'
+    expected_result = 5
+    result = ia._get_total_pages(start_timestamp, end_timestamp)
+    assert expected_result == result
```
I have changed the test in order to Mock API calls and use them for unit tests. Now the test is passing

amartya-dev commented:
@mathemancer I have made the requested changes. Please let me know if anything else is required. 

Also, just an idea, since it is up to us to decide how many rows we want per page, we can have multi-processing run different threads for pages having 24 rows each. Also, since we can parallelize the function, we can also include a write to disk option for each thread and thus, if the thread executes successfully we will be sure that that amount of rows have been written to the disk. Another way could be sharing the image_store object between threads so that we can know when the number of rows has reached a reasonable number in order to then hold further processing and write those changes to the disk.

I will try the above approach in case you approve of it until then I have made a commit corresponding to the changes requested.

mathemancer commented:
It might be worth exploring `asyncio`, since there's almost no chance this script is being slowed down by CPU.  I.e., the slow bit is the I/O over the network.  Perhaps using `asyncio`, along with `aiohttp` could speed things up marginally, if you make the requests for info about individual files concurrently.

I'd caution, however, that I suspect you won't get over a certain amount of throughput from the API, regardless of how many requests you make at once.  The reason is that they probably (almost certainly) have some rate-limiting implemented on their side that will simply slow down responses to large numbers of requests from the same IP.  In this case, hitting the endpoint with a bunch of requests in parallel (or concurrently in this case) will just irritate some sysadmin at Internet Archive without actually speeding anything up.

mathemancer reviewed (changes requested):
The only specific changes I'd request here would be to double-check for long lines in the file (`pycodestyle` is complaining).  Overall, this looks great, and I think it's gone as far as makes sense before finishing the DB, DB Loader, and data modelling work.

Thank you!

amartya-dev committed:
PEP8

amartya-dev commented:
@mathemancer Shortened the lines to comply with `pycodestyle`

mathemancer commented:
Closing until we're ready to work on audio.  Thank you for your effort!
